File: vm/src/types/object.c
/**
 * Copyright 2017 - Jahred Love
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice, this
 * list of conditions and the following disclaimer in the documentation and/or other
 * materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its contributors may
 * be used to endorse or promote products derived from this software without specific
 * prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS “AS IS” AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <jelly/internal.h>

#include <stdlib.h>
#include <string.h>

static uint32_t hash_u32(uint32_t x) {
  // A simple integer hash (xorshift/mix).
  x ^= x >> 16;
  x *= 0x7feb352du;
  x ^= x >> 15;
  x *= 0x846ca68bu;
  x ^= x >> 16;
  return x;
}

static void ensure_cap(jelly_object* o);

jelly_object* jelly_object_new(struct jelly_vm* vm, uint32_t type_id) {
  jelly_object* o = (jelly_object*)jelly_gc_alloc(vm, sizeof(jelly_object));
  o->h.kind = (uint32_t)JELLY_OBJ_OBJECT;
  o->h.type_id = type_id;
  o->proto = NULL;
  o->cap = 8;
  o->len = 0;
  o->keys = (uint32_t*)malloc(sizeof(uint32_t) * o->cap);
  o->vals = (jelly_value*)malloc(sizeof(jelly_value) * o->cap);
  o->states = (uint8_t*)malloc(o->cap);
  if(!o->keys || !o->vals || !o->states) abort();
  memset(o->states, 0, o->cap);
  return o;
}

static uint32_t find_slot(jelly_object* o, uint32_t atom_id, int* found) {
  uint32_t mask = o->cap - 1u;
  uint32_t h = hash_u32(atom_id);
  uint32_t i = h & mask;
  uint32_t first_tomb = UINT32_MAX;

  for(;;) {
    uint8_t st = o->states[i];
    if(st == 0) {
      *found = 0;
      return (first_tomb != UINT32_MAX) ? first_tomb : i;
    }
    if(st == 1 && o->keys[i] == atom_id) {
      *found = 1;
      return i;
    }
    if(st == 2 && first_tomb == UINT32_MAX) first_tomb = i;
    i = (i + 1u) & mask;
  }
}

int jelly_object_has(jelly_object* o, uint32_t atom_id) {
  int found = 0;
  (void)find_slot(o, atom_id, &found);
  return found;
}

jelly_value jelly_object_get(jelly_object* o, uint32_t atom_id) {
  int found = 0;
  uint32_t i = find_slot(o, atom_id, &found);
  if(!found) return jelly_make_null();
  return o->vals[i];
}

void jelly_object_set(jelly_object* o, uint32_t atom_id, jelly_value v) {
  ensure_cap(o);
  int found = 0;
  uint32_t i = find_slot(o, atom_id, &found);
  if(!found) {
    o->keys[i] = atom_id;
    o->states[i] = 1;
    o->len++;
  }
  o->vals[i] = v;
}

static void rehash(jelly_object* o, uint32_t new_cap) {
  uint32_t* old_keys = o->keys;
  jelly_value* old_vals = o->vals;
  uint8_t* old_states = o->states;
  uint32_t old_cap = o->cap;

  o->cap = new_cap;
  o->keys = (uint32_t*)malloc(sizeof(uint32_t) * o->cap);
  o->vals = (jelly_value*)malloc(sizeof(jelly_value) * o->cap);
  o->states = (uint8_t*)malloc(o->cap);
  if(!o->keys || !o->vals || !o->states) abort();
  memset(o->states, 0, o->cap);

  uint32_t old_len = o->len;
  o->len = 0;
  for(uint32_t i = 0; i < old_cap; i++) {
    if(old_states[i] == 1) {
      jelly_object_set(o, old_keys[i], old_vals[i]);
    }
  }
  o->len = old_len;

  free(old_keys);
  free(old_vals);
  free(old_states);
}

static void ensure_cap(jelly_object* o) {
  // Grow when load factor would exceed ~0.7
  if((o->len + 1u) * 10u < o->cap * 7u) return;
  rehash(o, o->cap * 2u);
}

File: jellyc/src/lower/expr/lit/object.rs
/*
 * Copyright 2022 - Jahred Love
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice, this
 * list of conditions and the following disclaimer in the documentation and/or other
 * materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its contributors may
 * be used to endorse or promote products derived from this software without specific
 * prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS “AS IS” AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

use crate::ast::ExprKind;
use crate::error::{CompileError, ErrorKind};
use crate::ir::{IrBuilder, IrOp, TypeId, VRegId};

use super::super::{intern_atom, is_object_kind, lookup_var, lower_expr, LowerCtx, T_OBJECT};

pub(super) fn lower_obj_lit(
    e: &crate::ast::Expr,
    fields: &[(String, crate::ast::Expr)],
    out_tid: TypeId,
    ctx: &mut LowerCtx,
    b: &mut IrBuilder,
) -> Result<(VRegId, TypeId), CompileError> {
    if out_tid != T_OBJECT && !is_object_kind(&ctx.type_ctx, out_tid) {
        return Err(CompileError::new(
            ErrorKind::Internal,
            e.span,
            "object literal has non-object semantic type",
        ));
    }
    let out = b.new_vreg(out_tid);
    b.emit(e.span, IrOp::ObjNew { dst: out });
    let proto_tid = ctx.proto_for_nominal;
    for (k, ve) in fields {
        let atom_id = intern_atom(k.as_str(), ctx);
        if let Some(nom_tid) = proto_tid {
            // Direct fn literal: recording_method will be consumed by fn_ lowering.
            ctx.recording_method = Some((nom_tid, atom_id));
            // Var ref: if binding has func_index, record now (before lower_expr overwrites env).
            if let ExprKind::Var(var_name) = &ve.node {
                if let Ok(bd) = lookup_var(ctx, var_name, e.span) {
                    if let Some(fi) = bd.func_index {
                        ctx.method_table.insert((nom_tid, atom_id), fi);
                    }
                }
            }
        }
        let (v_val, _t_val) = lower_expr(ve, ctx, b)?;
        ctx.recording_method = None;
        b.emit(
            e.span,
            IrOp::ObjSetAtom {
                obj: out,
                atom_id,
                value: v_val,
            },
        );
    }
    Ok((out, out_tid))
}
